flink-tests/src/test/java/org/apache/flink/test/recovery/BatchFineGrainedRecoveryITCase.java

```diff
@@ -110,13 +136,20 @@
 	private static final int MAX_JOB_RESTART_ATTEMPTS = ALL_MAPPERS_BACKTRACK_FAILURES + 2 * MAP_NUMBER;
 
 	/**
-	 * Expected restart number for each mapper.
-	 *
-	 * <p>Initial start plus exception failure plus restarts from this and each subsequent mapper's TM failure.
+	 * Expected attempt number for each mapper.
 	 */
-	private static final int[] EXPECTED_MAP_RESTARTS = IntStream
+	private static final int[] EXPECTED_MAP_ATTEMPT_NUMBERS = IntStream
 		.range(0, MAP_NUMBER)
-		.map(i -> 1 + 1 + MAP_NUMBER - i)
+		.map(i ->
+			// exception failure:
+				1 + // this mapper
+				i + // previous mappers
+			// TM failure:
+				(MAP_NUMBER - i - 1) + // subsequent mappers after PartitionNotFoundException
+				1 + // this mapper
+				1 + // this mapper after PartitionNotFoundException
+				i + // previous mappers
+				i) // previous mappers after PartitionNotFoundException
```

Comment:
Ok, we can try to simplify the test or you think this particular calculation could be rewritten?

Reply:
No I think the calculation itself with the comments is quite descriptive. However, the calculation itself is quite complicated because of all the different restarts being caused by different events.